### ISSUE TITLE
[ASViewController] Accessing vc.node.view doesn't go through ViewController's -loadView.

### ISSUE THREAD

**Issue by EugenePogrebnoy** [facebook/AsyncDisplayKit#2350]
Hi everybody,

I have a problem when trying to nest two `ASViewController`s: the nodes of both of them are loaded and laid out properly, but the `viewDidLoad()` method of the child view controller is not getting called. 

Here is my code:

```
import UIKit
import AsyncDisplayKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    var window: UIWindow?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        self.window = UIWindow()
        self.window?.rootViewController = ParentViewController()
        self.window?.makeKeyAndVisible()

        return true
    }
}

class ParentViewController: ASViewController<ASDisplayNode> {

    let child = ChildViewController()

    init() {
        super.init(node: ASDisplayNode())
        self.node.backgroundColor = .red

        self.node.usesImplicitHierarchyManagement = true
        self.node.layoutSpecBlock = { [child = self.child] _ in
            let childNode = child.node
            return ASInsetLayoutSpec(insets: UIEdgeInsetsMake(20, 20, 20, 20), child: childNode)
        }

        self.addChildViewController(self.child)
        self.child.didMove(toParentViewController: self)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.node.backgroundColor = .green
    }
}

class ChildViewController: ASViewController<ASDisplayNode> {
    init() {
        super.init(node: ASDisplayNode())
        self.node.backgroundColor = .blue
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.node.backgroundColor = .yellow
    }
}
```

![simulator screen shot 6 2016 12 53 51](https://cloud.githubusercontent.com/assets/430187/19148132/f67afaca-8bc3-11e6-9f2a-47ced969775a.png)

I expect the inner rectangle to be yellow.

Also according to the Apple docs the container view controller must call `didMove(toParentViewController:)` on its child after the transition to it which, as I understand, is after child view controller's node is loaded and inserted into the view hierarchy. What is the correct moment to call this method when using `ASDisplayNode`? Is there any way to detect that the node was inserted into the hierarchy without subclassing it?

I'm using AsyncDisplayKit version `1.9.90`

**Comment by mikhailmulyar** [comment]
I have the same issue with latest 2.0 version.

**Comment by hannahmbanana** [comment]
@EugenePogrebnoy @mikhailmulyar: Because you are using an `ASViewController` (subclass of a `UIViewController`), you should be able to use the viewWillAppear, etc. methods. Does this not work with the parent / child visibility call you mentioned?

**Comment by EugenePogrebnoy** [comment]
@hannahmbanana: Simply overriding `viewWillAppear()` and `viewDidAppear()` in the `ChildViewController` from the code above shows that they aren't called too.

I've taken a look at the `ASCellNode` implementation. What it is doing is explicitly calling `[_viewController view];`. The view controller then calls the `loadView()`, `viewDidLoad()` and all other methods.

This does the trick in my code too, but, in my opinion, it looks more like an error-prone workaround rather than a straightforward solution. I'd expect the `ASDisplayNode` to notify its view controller that it has loaded without requiring me to call anything explicitly.

I think the proper solution is to add an implicit relationship between an `ASDisplayNode` and its view controller. What do you think?

**Comment by appleguy** [comment]
@EugenePogrebnoy Thanks for the valuable description of the issue - I'd like to make sure this works correctly.

The right moment is almost certainly in -didEnterVisibleState and -didExitVisibleState.  For items in a scrolling container, this is called when at least one pixel of the cell is actually visible; for all other nodes, it's called when the item is added or removed from an active UIWindow sub-tree.

Although it would require a subclass, could you try this and let us know if it works?  If so, we could easily add a visibilityBlock or similar to be called on changes without subclassing.

If you have any issues with this, creating the analogous example for pure UIViewControllers that does work as you expect would be valuable in helping to illustrate the point of breakdown.

I'd be happy to assist in debugging this and ensure the right solution is upstreamed.  Please feel free to join us on Slack and correspond with me in a direct message as well!

**Comment by Adlai-Holler** [comment]
@EugenePogrebnoy Hi! I believe there are two aspects in play here:

There is the fact that `-[UIViewController view]` actually calls `-loadView` and `-viewDidLoad`, and so if you use `vc.node.view` to retrieve the view, that code will not be run. This is something we should think about how best to handle. We should also keep in mind that technically a user could pass a loaded node into `-initWithNode:`.

Separately, there is the fact that whenever you use child view controllers, it's important to add the view into the parent's view in between the calls to `addChildViewController` and `didMoveToParentViewController:`. So to fix the immediate issue, modifying as follows is a pretty standard convention for view controller containment:

```swift
class ParentViewController: ASViewController<ASDisplayNode> {

    let child = ChildViewController()

    init() {
        super.init(node: ASDisplayNode())
        self.node.backgroundColor = .red

        self.node.usesImplicitHierarchyManagement = true
        self.node.layoutSpecBlock = { [child = self.child] _ in
            let childNode = child.node
            return ASInsetLayoutSpec(insets: UIEdgeInsetsMake(20, 20, 20, 20), child: childNode)
        }
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.node.backgroundColor = .green

        self.addChildViewController(self.child)
        self.view.addSubview(self.child.view)
        self.child.didMove(toParentViewController: self)
    }
}
```

cc @appleguy We should discuss our options for handling the case where `-[UIViewController view]` is not called sometime soon.

**Comment by EugenePogrebnoy** [comment]
Sorry for not responding for so long.

@appleguy, I managed to solve my problem with the following class:
```swift
class ViewControllerContainerNode<ViewController : UIViewController>: ASDisplayNode {

    private(set) var viewController : ViewController?
    private var viewControllerBlock : (() -> ViewController)?
    private var viewControllerDidLoadBlock : ((ViewController) -> ())?
    
    weak var containerViewController : UIViewController? // Should be set explicitly by the node owner
    
    private var viewControllerNode : ASDisplayNode?
    
    init(viewController: ViewController) {
        self.viewController = viewController
        super.init()
        
        self.usesImplicitHierarchyManagement = true
    }
    
    init(viewControllerBlock: @escaping () -> ViewController, didLoad: ((ViewController) -> ())? = nil) {
        self.viewControllerBlock = viewControllerBlock
        self.viewControllerDidLoadBlock = didLoad
        
        super.init()
        
        self.usesImplicitHierarchyManagement = true
    }
    
    override func didLoad() {
        super.didLoad()
        
        guard let containerVC = self.containerViewController else {
            fatalError("ViewControllerContainerNode.containerViewController is not set")
        }
        
        if self.viewController == nil {
            self.viewController = self.viewControllerBlock?()
        }
        
        self.viewControllerBlock = nil
        
        guard let child = self.viewController else {
            fatalError("Failed to load child view controller")
        }
        
        containerVC.addChildViewController(child)
        
        _ = child.view // Load child's view if it hasn't been loaded

        if let child = child as? ASViewController<ASDisplayNode> {
            self.viewControllerNode = child.node
        }
        else {
            self.viewControllerNode = ASDisplayNode(viewBlock: { child.view }, didLoad: nil)
        }
        
        self.viewControllerDidLoadBlock?(child)
        self.viewControllerDidLoadBlock = nil
        
        guard let node = self.viewControllerNode else {
            fatalError("Failed to create view controller node")
        }
        
        node.layoutPosition = CGPoint.zero
        node.sizeRange = ASRelativeSizeRangeMakeWithExactPercent(1)
        
        self.setNeedsLayout()
    }
    
    private lazy var childDidMoveToParentDispatchOnce : () = {
        guard let containerVC = self.containerViewController else {
            fatalError("ViewControllerContainerNode.containerViewController is not set")
        }
        guard let child = self.viewController else {
            fatalError("Child should have already been initialized and loaded")
        }
        
        child.didMove(toParentViewController: containerVC)
    }()
    
    override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
        if let node = self.viewControllerNode {
            return ASStaticLayoutSpec(children: [node])
        }
        else {
            return ASStaticLayoutSpec(children: [])
        }
    }
    
    override func layoutDidFinish() {
        super.layoutDidFinish()
        
        guard self.viewControllerNode != nil else {
            return
        }
                
        _ = self.childDidMoveToParentDispatchOnce
    }
}
```

I tried it with both `UIViewController` and `ASViewController` subclasses and everything looks to work the right way.

Loading the view in response to the visibility events seems to work too, but then all loading and layout process happens immediately when the node becomes visible, which may cause some lags. So I thought that it's better to load the view as early as possible, i.e. in the `didLoad`.

Ideally, there should also be some logic to remove the child view controller from its parent when it's not needed anymore. That would be especially useful in the `ASCellNode` subclasses. Probably it should happen in response to interfaceState change events (`ASInterfaceStateMeasureLayout` or `ASInterfaceStateFetchData`). But I'm currently satisfied with the solution above.

It would also be nice to detect the correct container view controller automatically instead of setting it explicitly. But that might require some deep changes in `ASDiplayNode` class.


@Adlai-Holler, I generally agree with you. The only thing is that as I understand `self.view.addSubview(self.child.view)` is supposed to be called by ASDK when using `usesImplicitHierarchyManagement = true`. And while this approach solves the problem in the example, I'm afraid that it can become a nuisance in more complex node hierarchies.

**Comment by nguyenhuy** [comment]
@hannahmbanana: Looks like #2636 resolves this issue. Please assign this ticket to @maicki. Thanks!

**Comment by maicki** [comment]
#2636 was closed in favor of #2653. Due to the reason we decided against using the ViewControllers view as hosting view this issues still remains.

As @Adlai-Holler already mentioned I think we have to have in mind that the node hierarchy is different than the view hierarchy different than the view controller hierarchy. You always have to add the view within `addChildViewController` and `didMoveToParentViewController:`, so in theory `loadView` should be called with the right timing. Setting `automaticSubnodeManagement` to true just means that we manage the subnode hierarchy for you in this case.

The issue that `loadView` is not called for already loaded nodes still exists though.

**Comment by maicki** [comment]
Closing as should be fixed within #2656
